File: src/helius_agent/tools/git.py
import logging
import shlex
import subprocess
from pathlib import Path
from typing import List

from pydantic import BaseModel, Field
from langchain_core.tools import tool
from langgraph.types import interrupt

logger = logging.getLogger(__name__)

# --- Configuration & Security ---
WORKSPACE_ROOT = Path.cwd().resolve()
MAX_OUTPUT_LENGTH = 50_000

# Strict categorization of Git subcommands
SAFE_CMDS = {
    "status", "log", "diff", "show", "branch", "checkout", "switch", 
    "add", "commit", "stash", "pull", "fetch", "rev-parse"
}
DANGEROUS_CMDS = {
    "reset", "merge", "rebase", "clean", "restore", "revert", "rm"
}

def _run_git(args: List[str]) -> str:
    """
    Executes a git command securely.
    shell=False ensures no bash injections (e.g. `git commit -m "m" && rm -rf /`) can execute.
    """
    try:
        res = subprocess.run(
            ["git"] + args,
            cwd=str(WORKSPACE_ROOT),
            capture_output=True,
            text=True,
            check=True
        )
        output = res.stdout.strip() or "Success (no output)"
        return output[:MAX_OUTPUT_LENGTH] + ("\n...[TRUNCATED]" if len(output) > MAX_OUTPUT_LENGTH else "")
    except subprocess.CalledProcessError as e:
        error_msg = e.stderr.strip() or e.stdout.strip()
        return f"Git Error (Exit Code {e.returncode}):\n{error_msg}"


# ---------------------------------------------------------------------------
# 1. SAFE GIT TOOL
# ---------------------------------------------------------------------------
class GitSafeArgs(BaseModel):
    command: str = Field(
        ..., 
        description="The git command to execute (omitting the word 'git'). Examples: 'status --porcelain', 'commit -m \"fix\"', 'diff'."
    )

@tool(args_schema=GitSafeArgs)
def git_safe_op(command: str) -> str:
    """
    Execute a safe, non-destructive Git command (e.g., status, add, commit, branch).
    Do NOT use this for reset, merge, or force pushing.
    """
    try:
        args = shlex.split(command)
        if not args:
            return "Error: Empty command provided."

        subcmd = args[0]
        
        # Check for force push flags
        is_force_push = subcmd == "push" and any(f in args for f in ("--force", "-f", "--force-with-lease"))

        if subcmd in DANGEROUS_CMDS or is_force_push:
            return f"Security Error: '{subcmd}' operations are dangerous. You MUST use the `git_dangerous_op` tool instead."
            
        if subcmd not in SAFE_CMDS and subcmd != "push":
            return f"Security Error: Unrecognized or unauthorized safe command '{subcmd}'."

        return _run_git(args)
        
    except ValueError as e: # Catch shlex unmatched quotes
        return f"Argument Parsing Error: {str(e)}"
    except Exception as e:
        return f"Tool Execution Error: {str(e)}"


# ---------------------------------------------------------------------------
# 2. DANGEROUS GIT TOOL (Requires HITL)
# ---------------------------------------------------------------------------
class GitDangerousArgs(BaseModel):
    command: str = Field(
        ..., 
        description="The dangerous git command to execute. Examples: 'reset --hard HEAD~1', 'merge feature-branch'."
    )
    reason: str = Field(
        ..., 
        description="A clear explanation for the human of WHY this destructive action is needed."
    )

@tool(args_schema=GitDangerousArgs)
def git_dangerous_op(command: str, reason: str) -> str:
    """
    Execute a destructive or history-altering Git command (e.g., reset, merge, rebase, push --force).
    This tool pauses execution and requests human approval.
    """
    try:
        args = shlex.split(command)
        if not args:
            return "Error: Empty command provided."

        # Trigger Human-In-The-Loop (LangGraph v1)
        # This halts graph execution and yields this payload to the client.
        response = interrupt({
            "action": "approve_git",
            "command": f"git {command}",
            "reason": reason
        })

        # When the graph resumes, it injects the human's response here.
        if not response.get("approved", False):
            return f"Error: The human denied permission to run `git {command}`. Please find an alternative approach or ask for clarification."

        return _run_git(args)
        
    except ValueError as e:
        return f"Argument Parsing Error: {str(e)}"
    except Exception as e:
        return f"Tool Execution Error: {str(e)}"

# Bundle for the agent
GIT_TOOLS =[git_safe_op, git_dangerous_op]

File: source/tools/git.ts
import fs from 'node:fs';
import {spawnSync} from 'node:child_process';
import {randomUUID} from 'node:crypto';
import path from 'node:path';

import {tool} from 'langchain';
import {interrupt} from '@langchain/langgraph';
import {z} from 'zod';

import {getSettings} from '../utils/config.js';
import {getLogger} from '../utils/logger.js';
import {truncate} from '../utils/security.js';

const logger = getLogger('tools.git');

function workspaceRoot(): string {
	const root = getSettings().workspace.root;
	return path.resolve(root ?? process.cwd());
}

function splitArgs(command: string): string[] {
	const args: string[] = [];
	let current = '';
	let quote: '\'' | '"' | null = null;
	let escape = false;
	for (const ch of command) {
		if (escape) {
			current += ch;
			escape = false;
			continue;
		}
		if (ch === '\\' && quote !== '\'') {
			escape = true;
			continue;
		}
		if (ch === '\'' || ch === '"') {
			if (quote === ch) {
				quote = null;
			} else if (!quote) {
				quote = ch;
			} else {
				current += ch;
			}
			continue;
		}
		if (!quote && /\s/.test(ch)) {
			if (current.length > 0) {
				args.push(current);
				current = '';
			}
			continue;
		}
		current += ch;
	}
	if (current.length > 0) {
		args.push(current);
	}
	return args;
}

function runGit(args: string[]): string {
	try {
		logger.info(`git ${args.join(' ')}`);
		const result = spawnSync('git', args, {
			cwd: workspaceRoot(),
			encoding: 'utf8',
		});
		if (result.error) {
			return `Git error: ${result.error.message}`;
		}
		if (result.status !== 0) {
			const err = (result.stderr || result.stdout || '').trim();
			return `Git error (exit ${result.status}):\n${err}`;
		}
		return truncate((result.stdout || '').trim() || '(no output)', 50_000);
	} catch (err) {
		logger.error('git error', err);
		return `Git error: ${(err as Error).message}`;
	}
}

const gitSafeOp = tool(
	async ({command}: {command: string}) => {
		const cfg = getSettings().tools.git;
		let args: string[];
		try {
			args = splitArgs(command);
		} catch (err) {
			return `Argument parsing error: ${(err as Error).message}`;
		}
		if (args.length === 0) {
			return 'Error: Empty command.';
		}
		logger.info(`git_safe_op cmd=${command}`);

		const subcmd = args[0].toLowerCase();
		if (
			subcmd === 'push'
      && args.some(f => f === '--force' || f === '-f' || f === '--force-with-lease')
		) {
			return 'Error: Force-push is destructive. Use `git_dangerous_op` with a justification.';
		}
		if (cfg.dangerous_commands.has(subcmd)) {
			return `Error: '${subcmd}' is destructive. Use \`git_dangerous_op\` instead.`;
		}
		if (!cfg.safe_commands.has(subcmd) && subcmd !== 'push') {
			return `Error: Unrecognized git subcommand '${subcmd}'.`;
		}
		return runGit(args);
	},
	{
		name: 'git_safe_op',
		description: 'Execute a safe, non-destructive git command.',
		schema: z.object({
			command: z
				.string()
				.describe('Git subcommand and flags (omit the word \'git\').'),
		}),
	},
);

const gitDangerousOp = tool(
	async ({command, reason}: {command: string; reason: string}) => {
		let args: string[];
		try {
			args = splitArgs(command);
		} catch (err) {
			return `Argument parsing error: ${(err as Error).message}`;
		}
		if (args.length === 0) {
			return 'Error: Empty command.';
		}
		logger.info(`git_dangerous_op cmd=${command} reason=${reason}`);

		const response = interrupt({
			action: 'approve_git',
			command: `git ${command}`,
			reason,
		}) as Record<string, unknown>;

		if (!response?.approved) {
			logger.warn(`git_dangerous_op denied cmd=${command}`);
			return `Denied. Reason: ${String(response?.reason ?? 'None given')}. Find an alternative approach.`;
		}

		const edited = response?.edited_command as string | undefined;
		if (edited) {
			try {
				const cleaned = edited.startsWith('git ')
					? edited.slice(4)
					: edited;
				args = splitArgs(cleaned);
			} catch (err) {
				return `Argument parsing error in edited command: ${(err as Error).message}`;
			}
		}

		return runGit(args);
	},
	{
		name: 'git_dangerous_op',
		description: 'Execute a destructive git command (requires human approval).',
		schema: z.object({
			command: z
				.string()
				.describe('Destructive git command (without \'git\').'),
			reason: z
				.string()
				.describe('Clear justification for why this operation is necessary.'),
		}),
	},
);

const gitApplyPatchTool = tool(
	async ({patch}: {patch: string}) => {
		logger.info('git_apply_patch started');
		const ws = workspaceRoot();
		const tmpPath = path.join(ws, `.helius_tmp_${randomUUID().slice(0, 8)}.patch`);

		try {
			fs.writeFileSync(tmpPath, patch, 'utf8');
			const result = spawnSync('git', ['apply', '--whitespace=fix', tmpPath], {
				cwd: ws,
				encoding: 'utf8',
			});

			if (result.status !== 0) {
				const err = (result.stderr || result.stdout || 'Unknown error').trim();
				return `Git apply failed:\n${err}`;
			}

			return 'Patch applied successfully via git apply.';
		} catch (err) {
			logger.error('git_apply_patch error', err);
			return `Error: ${(err as Error).message}`;
		} finally {
			if (fs.existsSync(tmpPath)) {
				fs.unlinkSync(tmpPath);
			}
		}
	},
	{
		name: 'git_apply_patch',
		description: 'Apply a unified diff patch using \'git apply\'. More robust than manual patching.',
		schema: z.object({
			patch: z.string().describe('The unified diff patch to apply.'),
		}),
	},
);

export const GIT_TOOLS = [gitSafeOp, gitDangerousOp, gitApplyPatchTool];
